### PR TITLE
Docs / Code cleanup

### PR DESCRIPTION
**Pull request by kontsevoy**
Doing this PR a bit early, to break things apart a little bit. 

## Introduction

Big parts of this PR is documentation edits. You can ignore those. But I have touched the code as well. Some changes are cosmetic, like:

* Added license headers 
* Added code comments, pointing out at certain features that are soon to be deprecated
* Removed unused/obsolete facilities

Also, there are minor Teleport behavior changes.

## "Developer mode"

I discovered that `teleport` has a special "developer mode" which gets activated when `DEBUG` environment variable is set to `1`. A better name for this mode is probably "insecure developer mode" because when activated, TLS certificate validation is turned off.

I made two changes to it:

* I made it _explicit_. `DEBUG` is too generic (and quite common!) name, and someone may have this on by accident. To enable this behavior, a user must start the teleport daemon with `--insecure` flag now. 
* I made it easier to use and cheaper to maintain. It used to be implemented via regular configuration pipeline, making it incredibly expensive: several data structures and functions would carry this bit through, increasing the possibility of a bug. Moreover, it wasn't available in arbitrary location in the code (because the access to configuration structures is limited to certain components). Now this bit is set via simple `lib/runtimeflags.go` global variable and easily accessible from _any_ location in Teleport code, similar to how OSS/Enterprise switch is controlled.

## "Trustedcluster" role

Teleport has 3 built-in "service roles": proxy, node and auth. Servers with these roles can join a cluster by showing a token. A similar pattern was implemented for connecting clusters together, so a new "service role" was added called "trustedCluster". The name "Trustedcluster" gets directly exposed into `/etc/teleport.yaml` as `trustedcluster`, but everywhere else it's `trusted_cluster`. I also kept the old name around, calling it "legacy" in the code, for paranoia reasons.

So I renamed it to `trusted_cluster` for consistency.

## Bugfix

When connecting the trusted clusters together, _any error_ (including all network errors, for example when you misconfigure the port or the proxy address) would lead to "access denied" error message. 

## Error messages

I also ran into confusing situations (due to user error) where it wasn't clear why I was failing. I updated the errors to make it easier to understand.

### CODE REVIEW

**Comment by kontsevoy** [comment]
retest this please

File: docs/2.3/admin-guide.md [hunk 1]
```diff
@@ -326,71 +317,49 @@ proxy_service:
...
+  is managed by `tctl users` command. Teleport also supports second factor authentication 
+  (2FA) for the local connector. There are two types of 2FA:
     * [TOTP](https://en.wikipedia.org/wiki/Time-based_One-time_Password_Algorithm)
       is the default. You can use [Google Authenticator](https://en.wikipedia.org/wiki/Google_Authenticator) or 
       [Authy](https://www.authy.com/) or any other TOTP client.
     * [U2F](https://en.wikipedia.org/wiki/Universal_2nd_Factor) is the second.
-* **External users** are users stored elsewhere else within an organization. Examples include
-  Github, Active Directory (AD), LDAP server, OpenID/OAuth2 endpoint or behind SAML. 
+
+* **saml** connector type implements SAML authentication can be configured
```
Comment (klizhentas) [inline docs/2.3/admin-guide.md]: missing comma or conjunction?

**Comment by kontsevoy** [comment]
retest this please

File: lib/auth/trustedcluster.go [hunk 2]
```diff
@@ -180,6 +199,10 @@ func (a *AuthServer) establishTrust(trustedCluster services.TrustedCluster) erro
 	// send the request to the remote auth server via the proxy
 	validateResponse, err := a.sendValidateRequestToProxy(trustedCluster.GetProxyAddress(), &validateRequest)
 	if err != nil {
+		log.Error(err)
+		if strings.Contains(err.Error(), "x509") {
```
Comment (klizhentas) [inline lib/auth/trustedcluster.go]: I think you can use type casting here that is more robust than string submatch:

```
case x509.SystemRootsError, x509.UnknownAuthorityError:
```

Comment (kontsevoy) [reply lib/auth/trustedcluster.go]: I thought so too. Unfortunately that's not an error from the http package, it's our own JSON error from the API.

File: lib/runtimeflags.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	currentDistroType teleport.DistroType = teleport.DistroTypeOSS
+
+	// insecureDevMode is set to 'true' when teleport is started with a hidden
+	// --insecure flag. This mode is only useful for learning Teleport and following
+	// quick starts: it disables HTTPS certificate validation
+	insecureDevMode bool
+)
+
+func SetDistroType(t teleport.DistroType) {
+	currentDistroType = t
```
Comment (klizhentas) [inline lib/runtimeflags.go]: you should protect this by locks, as read write from global vairable is not a thread safe operation.

File: lib/auth/trustedcluster.go (hunk 2 above)
Comment (kontsevoy) [reply lib/auth/trustedcluster.go]: `trace.IsTrustError()` fails on it too.

File: lib/runtimeflags.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+func IsEnterprise() bool {
```
Comment (klizhentas) [inline lib/runtimeflags.go]: missing comment of an exported method

File: lib/runtimeflags.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	return currentDistroType == teleport.DistroTypeEnterprise
+}
+
+func SetInsecureDevMode(m bool) {
```
Comment (klizhentas) [inline lib/runtimeflags.go]: should be protected by lock

**Review by klizhentas — CHANGES_REQUESTED** [review]
need to protect global variable by lock,

**Review by klizhentas — APPROVED** [review]

**Pull request merged**